docs/design/001-prometheus.md

```diff
@@ -0,0 +1,445 @@
+
+The prometheus support should be supported in both cases, when running in Kubernetes (ig-k8s), or
+when running on Linux hosts (ig). This is possible by implementing this using a new Prometheus
+gadget/operator as it makes the code automatically shareable between ig, ig-k8s and external
+applications. This gadget /operator provides start / stop operations to enable / disable this
+support.
+
+```bash
+$ kubectl gadget prometheus start --config <path>
+$ kubectl gadget prometheus stop
+```
+
+TODO: need to think about not having a start operation on this one
+
+```bash
+$ ig prometheus --config <path>
+```
+
+This is also possible to configure the metrics using a CR and supporting a [static
+configuration](https://github.com/inspektor-gadget/inspektor-gadget/issues/1401) could be
+implemented in the future as well.
+
+## Configuration File
+
+Given that we want this support to be flexible allowing the user to control which metrics to capture
+and how to aggregate them. The configuration file defines those aspects. It takes inspiration from
+[https://github.com/cloudflare/ebpf_exporter](https://github.com/cloudflare/ebpf_exporter).
+
+### Filtering (aka Selectors)
+
+The user should be able to provide a set of filters indicating the events that should be taken into
+consideration when updating the metrics. The mechanism should provide the following features:
+
+- Equals operator (==)
+- Different operator (!=)
+- Greater than, less than operators (<, >, <=, >=)
+- Set based
+  - In
+  - NotIn
+
+It's similar to the existing [Labels and
+Selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) mechanism, but
+we still need to understand if we can reuse that or if we need a completely new implementation.
+
+TODO: we still need to understand which operator we really need.
+
+Some examples of possible filters are:
+
+# Only metrics for default namespace
+
+```yaml
+# Only metrics for default namespace
+selector:
+  - k8s.namespace: default
+
+# Count only events with retval != 0
+# TODO: still needed to understand the syntax for the != operator
+selector:
+  matchExpressions:
+  - { key: retval, operator: NotIn, values: [0] }
+```
+
+The configuration file defines the different metrics to collect:
+
+- counters
+- gauges
+- histograms
+
+### Counters
+
+This is probably the most intuitive metric: "A _counter_ is a cumulative metric that represents a
+single [monotonically increasing counter](https://en.wikipedia.org/wiki/Monotonic_function) whose
+value can only increase or be reset to zero on restart. For example, you can use a counter to
+represent the number of requests served, tasks completed, or errors." from
+[https://prometheus.io/docs/concepts/metric_types/#counter](https://prometheus.io/docs/concepts/metric_types/#counter).
+
+The following are examples of counters we can support with the existing gadgets. The first one
+counts the number of executed processes.
+
+```yaml
+metrics:
+  counters:
```

Comment:
IMHO, it would be better to have these keys (counters, gauges, histograms) removed and instead use a `type` field inside the actual metric configuration with a value of counter, gauge and histogram respectively. I know that the proposed change can be a bit trickier to parse, but from a user perspective it seems better and cleaner to be able to group metrics that belong together, regardless of their type.

Reply:
Makes sense. I updated the document to follow this approach.